{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "import os\n",
    "directory = '../../output.ibd2'\n",
    "render='svg'\n",
    "name=os.path.basename(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from charcoal.utils import CSV_DictHelper\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "summary = CSV_DictHelper(f'{directory}/stage1_genome_summary.csv', 'genome')\n",
    "display(Markdown(f\"# charcoal summary report for `{name}`\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from charcoal import utils\n",
    "import collections\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate and display some basic statistics\n",
    "all_genomes = set(summary)\n",
    "\n",
    "n_genomes = len(all_genomes)\n",
    "no_lineage = set([ k for k in all_genomes if not summary[k].lineage ])\n",
    "n_no_lineage = len(no_lineage)\n",
    "all_genomes -= no_lineage\n",
    "\n",
    "no_order_contam = set([ k for k in all_genomes if int(summary[k].bad_order_bp) == 0 ])\n",
    "n_no_contam = len(no_order_contam)\n",
    "all_genomes -= no_order_contam\n",
    "contaminated_list = set(all_genomes)\n",
    "clean_list = set(no_order_contam)\n",
    "\n",
    "needs_lineage = [ k for k in no_lineage if int(summary[k].needs_lineage_flag) ]\n",
    "n_needs_lineage = len(needs_lineage)\n",
    "\n",
    "first_contam = {}\n",
    "ignore = set()\n",
    "\n",
    "for rank in 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus':\n",
    "    n = 0\n",
    "    for genome in all_genomes:\n",
    "        bad_bp = int(summary[genome][f'bad_{rank}_bp'])\n",
    "        if genome in ignore:\n",
    "            continue\n",
    "        if bad_bp:\n",
    "            n += 1\n",
    "            ignore.add(genome)\n",
    "            continue\n",
    "    first_contam[rank] = n\n",
    "    \n",
    "def display_genome_list(names_to_show, order_by=None, reverse=True):\n",
    "    # select only those we want to show\n",
    "    values = [ summary[k] for k in names_to_show ]\n",
    "    \n",
    "    # sort by:\n",
    "    if order_by:\n",
    "        values.sort(key=lambda x: int(x[order_by]))\n",
    "        if reverse: values.reverse()\n",
    "    else:\n",
    "        values.sort(key = lambda x: x.genome)\n",
    "    \n",
    "    output = []\n",
    "    for row in values:\n",
    "        output.append(f\"* `{row.genome}` [taxonomic report]({row.genome}.fig.html) - [alignment report]({row.genome}.align.html)\")\n",
    "    \n",
    "    return \"\\n\".join(output)\n",
    "\n",
    "def display_no_lineage_list():\n",
    "    values = [ summary[k] for k in summary if not summary[k].lineage ]\n",
    "    \n",
    "    x = []\n",
    "    for row in values:\n",
    "        x.append(f\"* `{row.genome}` - {row.comment}\")\n",
    "        \n",
    "    return \"\\n\".join(x)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = [\"all genomes\", \"all analyzed\", \"contaminated\", \"no contamination\", \"not analyzed\", \"no matches\", \"needs lineage\"]\n",
    "parents = [ \"\", \"all genomes\", \"all analyzed\", \"all analyzed\", \"all genomes\", \"not analyzed\", \"not analyzed\"]\n",
    "\n",
    "# get basic numbers in --\n",
    "values = [ len(summary),\n",
    "           len(summary) - n_no_lineage,\n",
    "           len(contaminated_list),\n",
    "           n_no_contam,\n",
    "           n_no_lineage,\n",
    "           n_no_lineage - n_needs_lineage,\n",
    "           n_needs_lineage,\n",
    "           ]\n",
    "\n",
    "# add per-rank breakdown\n",
    "for rank in reversed(['superkingdom', 'phylum', 'class', 'order', 'family', 'genus']):\n",
    "    labels.append(rank)\n",
    "    parents.append('contaminated')\n",
    "    values.append(first_contam[rank])\n",
    "    \n",
    "# display!\n",
    "\n",
    "fig = go.Figure(go.Treemap(\n",
    "    branchvalues = \"total\",\n",
    "    labels = labels,\n",
    "    parents = parents,\n",
    "    values = values,\n",
    "    textinfo = \"label+value+percent entry\",\n",
    "    outsidetextfont = {\"size\": 20, \"color\": \"darkblue\"},\n",
    "    marker = {\"line\": {\"width\": 2}},\n",
    "    pathbar = {\"visible\": False}))\n",
    "\n",
    "fig.update_layout(treemapcolorway = [\"lavenderblush\", \"lightblue\"],\n",
    "                  title=f'summary of charcoal run for {name}')\n",
    "fig.show(render, width=1000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### do display\n",
    "\n",
    "display(Markdown(f\"\"\"\n",
    "# Summary statistics\n",
    "\n",
    "charcoal examined {n_genomes} total genomes in `{name}`.\n",
    "\n",
    "{n_no_lineage} could not be analyzed for contamination; see Problem Report, below.\n",
    "\n",
    "{n_no_contam} had no observed contamination at any taxonomic level down to genus.\n",
    "\n",
    "Of the remaining {len(contaminated_list)} genomes,\n",
    "* {first_contam['superkingdom']} had their first observed contamination at **superkingdom**;\n",
    "* {first_contam['phylum']} had their first observed contamination at **phylum**;\n",
    "* {first_contam['class']} had their first observed contamination at **class**;\n",
    "* {first_contam['order']} had their first observed contamination at **order**;\n",
    "* {first_contam['family']} had their first observed contamination at **family**;\n",
    "* {first_contam['genus']} had their first observed contamination at **genus**.\n",
    "\"\"\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_hitlist_data(hit_list_filename, omit_genus=True, num=10):\n",
    "    hits = utils.CSV_DictHelper(hit_list_filename, 'genome')\n",
    "    \n",
    "    endings = '.fna.gz', 'fa.gz', '.fa', '_genomic'\n",
    "    \n",
    "    # sort by sum of bad bp at specific rank\n",
    "    if omit_genus:\n",
    "        sort_by = 'family_bad_bp'\n",
    "    else:\n",
    "        sort_by = 'genus_bad_bp'\n",
    "    \n",
    "    # pick off the top num items, sorted by given sort_by\n",
    "    items = list(hits.rows.items())\n",
    "    items.sort(key=lambda item: -int(item[1][sort_by]))\n",
    "    items = items[:num]\n",
    "    items.reverse()\n",
    "    \n",
    "    hitlist_data = collections.defaultdict(list)\n",
    "    names = hitlist_data['names']\n",
    "    \n",
    "    # for each item, retrieve all tax rank bp values\n",
    "    for name, row in items:\n",
    "        # clean up names:\n",
    "        for end in endings:\n",
    "            if name.endswith(end): name = name[:-len(end)]\n",
    "        names.append(name)\n",
    "\n",
    "        last_bp = 0\n",
    "        for key in 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus':\n",
    "            bp = row[f'{key}_bad_bp']\n",
    "            bp = int(bp)\n",
    "            values = hitlist_data[key]\n",
    "            values.append(bp - last_bp)\n",
    "            last_bp = bp\n",
    "            \n",
    "    return hitlist_data\n",
    "\n",
    "def plot_hitlist(hit_list_filename, omit_genus=True, num=10):\n",
    "    hitlist_data = make_hitlist_data(hit_list_filename, omit_genus, num)\n",
    "    \n",
    "    names = hitlist_data['names']\n",
    "\n",
    "    barchart_objs = []\n",
    "    for key in 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus':\n",
    "        values = hitlist_data[key]\n",
    "        if omit_genus and key == 'genus': break\n",
    "            \n",
    "        bc_obj = go.Bar(name=key, y=names, x=values, orientation='h')\n",
    "        barchart_objs.append(bc_obj)\n",
    "        \n",
    "    fig = go.Figure(data=barchart_objs)\n",
    "    title = f'top genome contamination by taxonomic rank: {os.path.dirname(hit_list_filename)[7:]}'\n",
    "    fig.update_layout(barmode='stack', template='plotly',\n",
    "                      xaxis_title='Estimated contamination by rank',\n",
    "                      title=title)\n",
    "\n",
    "    return fig\n",
    "\n",
    "fig = plot_hitlist(f'{directory}/stage1_hitlist.csv', omit_genus=False)\n",
    "if not render: render_name = None\n",
    "else: render_name = render\n",
    "fig.show(render_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(f\"\"\"\n",
    "## Individual contamination reports\n",
    "\n",
    "(Sorted by summed contamination at order level.)\n",
    "\n",
    "{display_genome_list(contaminated_list, 'bad_order_bp')}\n",
    "\n",
    "## Problem Report\n",
    "\n",
    "{n_no_lineage} genomes could not be analyzed for contamination.\n",
    "\n",
    "{display_no_lineage_list()}\n",
    "\n",
    "## Analyzed genomes for which no removable contamination was detected\n",
    "\n",
    "{n_no_contam} genomes had no removable contamination at any taxonomic level down to genus.\n",
    "\n",
    "(Some of this could be due to cryptic contamination or chimeric contigs. @CTB document.)\n",
    "\n",
    "{display_genome_list(clean_list)}\n",
    "\n",
    "\"\"\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (myenv)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
