{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# example parameters\n",
    "directory = '../../output.nigoni'\n",
    "#name = 'SRS104400_110.fna.gz'\n",
    "name = 'nigoni_canu_2018.09.14.01.fa.gz'\n",
    "render='svg'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not render: render_name = None\n",
    "else: render_name = render"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os.path\n",
    "import collections\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "import sourmash\n",
    "from charcoal import utils\n",
    "from charcoal.figs.sourmash_sankey import GenomeSankeyFlow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown as md\n",
    "from IPython.display import display\n",
    "md(f\"# Charcoal summary report for genome `{name}`\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from charcoal import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_csv = f'{directory}/stage1_genome_summary.csv'\n",
    "summary = utils.CSV_DictHelper(summary_csv, 'genome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "info = summary[name]\n",
    "\n",
    "def bp(num):\n",
    "    num = int(num)\n",
    "    if num > 2e6:\n",
    "        return f\"{num / 1e6:.1f} Mbp\"\n",
    "    elif num > 2e3:\n",
    "        return f\"{num / 1e3:.1f} kb\"\n",
    "    else:\n",
    "        return f\"{num} bp\"\n",
    "\n",
    "x = []\n",
    "\n",
    "lineage = info['lineage'].strip()\n",
    "no_lineage = False\n",
    "if not lineage:\n",
    "    lineage = '*none available*'\n",
    "    no_lineage = True\n",
    "x.append(f\" Genome lineage: `{lineage}`\")\n",
    "display(md(\"\\n\".join(x)))\n",
    "\n",
    "if no_lineage:\n",
    "    raise SystemExit(0)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []\n",
    "\n",
    "x.append(\"### estimated clean / contaminated bp at each taxonomic rank\")\n",
    "for rank in ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus']:\n",
    "    bad_name = f'bad_{rank}_bp'\n",
    "    good_name = f'good_{rank}_bp'\n",
    "\n",
    "    x.append(f\"* {rank}: good {bp(info[good_name])} / bad {bp(info[bad_name])}\")\n",
    "\n",
    "x.append(f\"\"\"\n",
    "\n",
    "Of the total {bp(info['total_contigs_bp'])} in the genome file,\n",
    "{bp(info['noident_contigs_bp'])} could not be identified with the database used, and\n",
    "{bp(info['ignored_contigs_bp'])} was ignored by sourmash due to the scaled/ksize values used.\n",
    "\"\"\")\n",
    "    \n",
    "    \n",
    "display(md(\"\\n\".join(x)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How clean and dirty content changes by taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genome_area_chart(genome, summary):\n",
    "    row = summary[genome]\n",
    "\n",
    "    x=['superkingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "\n",
    "    y_ignored_bp = []\n",
    "    y_noident_bp = []\n",
    "    y_clean_bp = []\n",
    "    y_dirty_bp = []\n",
    "\n",
    "    ignored_bp = int(row['ignored_contigs_bp'])\n",
    "    noident_bp = int(row['noident_contigs_bp'])\n",
    "\n",
    "    for rank in x:\n",
    "        # adjust good_bp for unidentified bp and ignored bp.\n",
    "        tax_good_bp = int(row[f'good_{rank}_bp']) - noident_bp - ignored_bp\n",
    "        tax_bad_bp = int(row[f'bad_{rank}_bp'])\n",
    "        # double check calculations, 'cause why not?\n",
    "        assert ignored_bp + noident_bp + tax_good_bp + tax_bad_bp == int(row['total_contigs_bp'])\n",
    "\n",
    "        y_ignored_bp.append(ignored_bp)\n",
    "        y_noident_bp.append(noident_bp)\n",
    "        y_clean_bp.append(tax_good_bp)\n",
    "        y_dirty_bp.append(tax_bad_bp)\n",
    "\n",
    "    fig = go.Figure()\n",
    "\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=x, y=y_clean_bp,\n",
    "        mode='lines',\n",
    "        line=dict(width=0.5, color='rgb(184, 247, 212)'),\n",
    "        stackgroup='one',\n",
    "        name='clean bp',\n",
    "        groupnorm='percent' # sets the normalization for the sum of the stackgroup\n",
    "    ))\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=x, y=y_noident_bp,\n",
    "        mode='lines',\n",
    "        line=dict(width=0.5, color='rgb(111, 231, 219)'),\n",
    "        stackgroup='one',\n",
    "        name='no ident'\n",
    "    ))\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=x, y=y_ignored_bp,\n",
    "        mode='lines',\n",
    "        line=dict(width=0.5, color='rgb(127, 166, 238)'),\n",
    "        stackgroup='one',\n",
    "        name='no hashes (ignored)'\n",
    "    ))\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=x, y=y_dirty_bp,\n",
    "        mode='lines',\n",
    "        line=dict(width=0.5, color='rgb(131, 90, 241)'),\n",
    "        stackgroup='one',\n",
    "        name='dirty bp'\n",
    "    ))\n",
    "\n",
    "    fig.update_layout(\n",
    "        showlegend=True,\n",
    "        xaxis_type='category',\n",
    "        yaxis=dict(\n",
    "            type='linear',\n",
    "            range=[1, 100],\n",
    "            ticksuffix='%'))\n",
    "    \n",
    "    title = genome\n",
    "    fig.update_layout(template='plotly',\n",
    "                      title=title)\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig = plot_genome_area_chart(name, summary)\n",
    "fig.show(render_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Taxonomic assignments across ranks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_plotly_fig(obj, title):\n",
    "    \"Build a plotly figure/sankey diagram.\"\n",
    "    # make the data to go into the sankey figure.\n",
    "    labels = obj.make_labels()\n",
    "    src_l, dest_l, cnt_l, color_l, label_l = obj.make_lists()\n",
    "\n",
    "    # build figure\n",
    "    fig = go.Figure(data=[go.Sankey(\n",
    "        node = dict(\n",
    "          pad = 15,\n",
    "          thickness = 20,\n",
    "          line = dict(color = \"black\", width = 0.5),\n",
    "          label = labels,\n",
    "          color = \"blue\"\n",
    "        ),\n",
    "        link = dict(\n",
    "          source = src_l,\n",
    "          target = dest_l,\n",
    "          value = cnt_l,\n",
    "          color = color_l,\n",
    "          label = label_l,\n",
    "      ))])\n",
    "\n",
    "    if title:\n",
    "        fig.update_layout(title_text=title, font_size=10)\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_plotly_fig(obj, title):\n",
    "    \"Build a plotly figure/sankey diagram.\"\n",
    "    # make the data to go into the sankey figure.\n",
    "    labels = obj.make_labels()\n",
    "    src_l, dest_l, cnt_l, color_l, label_l = obj.make_lists()\n",
    "\n",
    "    # build figure\n",
    "    fig = go.Figure(data=[go.Sankey(\n",
    "        node = dict(\n",
    "          pad = 15,\n",
    "          thickness = 20,\n",
    "          line = dict(color = \"black\", width = 0.5),\n",
    "          label = labels,\n",
    "          color = \"blue\"\n",
    "        ),\n",
    "        link = dict(\n",
    "          source = src_l,\n",
    "          target = dest_l,\n",
    "          value = cnt_l,\n",
    "          color = color_l,\n",
    "          label = label_l,\n",
    "      ))])\n",
    "\n",
    "    if title:\n",
    "        fig.update_layout(title_text=title, font_size=10)\n",
    "\n",
    "    return fig\n",
    "\n",
    "\n",
    "def make_flow_fig(title, genome_lineage, contigs_info):\n",
    "    obj = GenomeSankeyFlow()\n",
    "    \n",
    "    counts = collections.Counter()\n",
    "    for contig_name, gather_info in contigs_info.items():\n",
    "        contig_taxlist = gather_info.gather_tax\n",
    "\n",
    "        # iterate over each contig match and summarize counts.              \n",
    "        # note - here we can stop at first one, or track them all.          \n",
    "        # note - b/c gather counts each hash only once, these               \n",
    "        #     are non-overlapping                                           \n",
    "        total_hashcount = 0\n",
    "        for lin, hashcount in contig_taxlist:\n",
    "            counts[lin] += hashcount\n",
    "            total_hashcount += hashcount\n",
    "            break # display only first\n",
    "\n",
    "        # track missing => unassigned lineage\n",
    "        unident = gather_info.num_hashes - total_hashcount\n",
    "        counts[obj.unassigned_lin] += unident\n",
    "\n",
    "    # extend genome_lineage ranks\n",
    "    last_genome_rank = genome_lineage[-1].rank\n",
    "    taxlist = list(sourmash.lca.taxlist())[:-2]\n",
    "    assert taxlist[-1] == 'genus', taxlist\n",
    "    \n",
    "    if genome_lineage[-1].rank != 'genus':\n",
    "        genome_lineage = list(genome_lineage)\n",
    "        for rank in taxlist[len(genome_lineage):]:\n",
    "            genome_lineage.append(sourmash.lca.LineagePair(rank=rank, name='unassigned'))\n",
    "    \n",
    "    # set the color of the main lineage\n",
    "    genome_lineage = tuple(genome_lineage)\n",
    "    obj.colors[genome_lineage] = \"lightseagreen\"\n",
    "    \n",
    "    # for phylum level disagreements, let's go with palevioletred\n",
    "    for lin in counts:\n",
    "        if not utils.is_lineage_match(lin, genome_lineage, 'phylum'):\n",
    "            obj.colors[lin] = 'palevioletred'\n",
    "            \n",
    "    # assign unassigned to good lineage, maybe?\n",
    "    counts[genome_lineage] += counts[obj.unassigned_lin]\n",
    "    del counts[obj.unassigned_lin]\n",
    "    \n",
    "    obj.make_links(genome_lineage, counts)\n",
    "    fig = make_plotly_fig(obj, title)\n",
    "    \n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs_filename = f'{directory}/stage1/{name}.contigs-tax.json'\n",
    "\n",
    "summary = utils.CSV_DictHelper(summary_csv, 'genome')\n",
    "row = summary[name]\n",
    "genome_lineage = utils.make_lineage(row.lineage)\n",
    "\n",
    "contigs_info = utils.load_contigs_gather_json(contigs_filename)\n",
    "\n",
    "title = sourmash.lca.display_lineage(genome_lineage)\n",
    "fig = make_flow_fig(title, genome_lineage, contigs_info)\n",
    "fig.show(render_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (myenv)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
